report/firefox_code_coverage/codecoverage.py

```diff
@@ -28,91 +24,82 @@
 GRCOV_ARTIFACT = "public/build/grcov.tar.xz"
 
 
-def get_json(url, params=None, headers={}):
-    if params is not None:
-        url += "?" + urlencode(params)
-
-    request = Request(url, headers=headers)
-    r = urlopen(request).read().decode("utf-8")
-
-    return json.loads(r)
-
-
 def is_taskcluster_loaner():
     return "TASKCLUSTER_INTERACTIVE" in os.environ
 
 
 def get_task(branch, revision):
-    task = get_json(
-        f"https://firefox-ci-tc.services.mozilla.com/api/index/v1/task/gecko.v2.{branch}.revision.{revision}.firefox.decision"
-    )
+    index = taskcluster.get_service("index")
+    task = index.findTask(f"gecko.v2.{branch}.revision.{revision}.firefox.decision")
     return task["taskId"]
 
 
 def get_last_task():
-    revision = get_json(
+    resp = requests.get(
         "https://api.coverage.moz.tools/v2/latest?repository=mozilla-central"
-    )[0]["revision"]
+    )
+    resp.raise_for_status()
+    data = resp.json()
+    revision = data[0]["revision"]
     return get_task("mozilla-central", revision)
 
 
 def get_task_details(task_id):
-    task_details = get_json(
-        "https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/" + task_id
-    )
-    return task_details
+    queue = taskcluster.get_service("queue")
+    return queue.task(task_id)
 
 
 def get_task_artifacts(task_id):
-    artifacts = get_json(
-        "https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/"
-        + task_id
-        + "/artifacts"
-    )
-    return artifacts["artifacts"]
+    queue = taskcluster.get_service("queue")
+    response = queue.listLatestArtifacts(task_id)
+    return response["artifacts"]
 
 
 def get_tasks_in_group(group_id):
-    reply = get_json(
-        "https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task-group/"
-        + group_id
-        + "/list",
-        {"limit": "200"},
-    )
-    tasks = reply["tasks"]
-    while "continuationToken" in reply:
-        reply = get_json(
-            "https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task-group/"
-            + group_id
-            + "/list",
-            {"limit": "200", "continuationToken": reply["continuationToken"]},
-        )
-        tasks += reply["tasks"]
+    tasks = []
+
+    def _save_tasks(response):
+        tasks.extend(response["tasks"])
+
+    queue = taskcluster.get_service("queue")
+    queue.listTaskGroup(group_id, paginationHandler=_save_tasks)
+
     return tasks
 
 
+def download_binary(url, path):
+    """Download a binary file from an url"""
+    while True:
+        try:
+            artifact = requests.get(url, stream=True)
+            artifact.raise_for_status()
+
+            with open(path, "wb") as f:
+                for chunk in artifact.iter_content(chunk_size=8192):
+                    f.write(chunk)
+            break
+        except:  # noqa: E722
+            try:
+                os.remove(path)
+            except OSError:
+                pass
+
+            time.sleep(7)
```

Comment:
We can potentially spin for ever..., so I'd run this for max_retries and use a backoff

Reply:
Done !